### PR TITLE
Bypass LIMS for delivery report

### PR DESCRIPTION
**Pull request by ivadym**
## Description

The goal is to avoid the delivery report crashing due to samples not being present in LIMS (HTTP exception). While the report creation will still fail, it'll be possible to work around it with the `--force flag`, including empty data for the application info, methods, and metrics for Rnafusion.

I've also moved some methods from the report api to the analysis api since I've seen some duplication that could be introduced here: https://github.com/Clinical-Genomics/cg/pull/3112

### Added
- Bypass LIMS for delivery report

### Changed
- Moved analysis specific methods from the report API to analysis

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### Tests

- [x] Generate delivery report for a sample that's not in LIMS:
    - [x]  With the current master branch (exception raised and not captured, not possible to bypass generation with `--force`):

    - [x] With this branch (generation fails gracefully, logging the missing report data. `--force` flag generates a report with empty data):

- [x] `Balsamic` delivery-report
- [x] `Rnafusion` delivery-report
- [x] `MIP-DNA` delivery-report

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/apps/lims/api.py [hunk 1]
```diff
@@ -53,10 +54,15 @@ def __init__(self, config):
     def user(self) -> Researcher:
         return self.get_researchers(username=self.username)[0]
 
-    def sample(self, lims_id: str):
+    def sample(self, lims_id: str) -> dict[str, Any]:
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:57]:
```suggestion
    def get_sample(self, lims_id: str) -> dict[str, Any]:
```
Disregard if out of scope.

File: cg/apps/lims/api.py [hunk 2]
```diff
@@ -53,10 +54,15 @@ def __init__(self, config):
...
         """Fetch a sample from the LIMS database."""
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:58]:
```suggestion
        """Return a sample from the LIMS database."""
```

File: cg/apps/lims/api.py [hunk 3]
```diff
@@ -128,35 +134,31 @@ def get_delivery_date(self, lims_id: str) -> dt.date:
             date = None
         return date
 
-    def capture_kit(self, lims_id: str) -> str:
+    def capture_kit(self, lims_id: str) -> str | None:
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:137]:
```suggestion
    def get_capture_kit(self, lims_id: str) -> str | None:
```

File: cg/apps/lims/api.py [hunk 4]
```diff
@@ -128,35 +134,31 @@ def get_delivery_date(self, lims_id: str) -> dt.date:
...
         """Get capture kit for a LIMS sample."""
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:138]:
```suggestion
        """Return capture kit for a LIMS sample."""
```

File: cg/apps/lims/api.py [hunk 5]
```diff
@@ -128,35 +134,31 @@ def get_delivery_date(self, lims_id: str) -> dt.date:
...
-        if len(capture_kits) > 1:
-            message = f"Capture kit error: {lims_sample.id} | {capture_kits}"
-            raise LimsDataError(message)
-
-        if len(capture_kits) == 1:
-            return capture_kits.pop()
-
+        try:
+            lims_sample = Sample(self, id=lims_id)
+            capture_kit = lims_sample.udf.get("Bait Set")
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:143]: Would be great with a constant class instead of hard coded keys.
Add type hint.

File: cg/apps/lims/api.py [hunk 6]
```diff
@@ -128,35 +134,31 @@ def get_delivery_date(self, lims_id: str) -> dt.date:
...
+            if capture_kit and capture_kit != "NA":
+                return capture_kit
+            for process_type in step_names_udfs:
+                artifacts = self.get_artifacts(
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:147]: Add type hint

File: cg/apps/lims/api.py [hunk 7]
```diff
@@ -128,35 +134,31 @@ def get_delivery_date(self, lims_id: str) -> dt.date:
...
+                    samplelimsid=lims_id, process_type=process_type, type="Analyte"
+                )
+                udf_key = step_names_udfs[process_type]
+                capture_kits = capture_kits.union(
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:151]: Add type hint

File: cg/apps/lims/api.py [hunk 8]
```diff
@@ -128,35 +134,31 @@ def get_delivery_date(self, lims_id: str) -> dt.date:
...
+                    self._find_capture_kits(artifacts, udf_key)
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:152]: Use named args

File: cg/apps/lims/api.py [hunk 9]
```diff
@@ -239,19 +241,17 @@ def get_sample_attribute(self, lims_id: str, key: str) -> str:
             )
         return sample.udf[PROP2UDF[key]]
 
-    def get_prep_method(self, lims_id: str) -> str:
+    def get_prep_method(self, lims_id: str) -> str | None:
         """Get the library preparation method."""
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:245]:
```suggestion
        """Return the library preparation method."""
```

File: cg/apps/lims/api.py [hunk 10]
```diff
@@ -239,19 +241,17 @@ def get_sample_attribute(self, lims_id: str, key: str) -> str:
...
-
         step_names_udfs = MASTER_STEPS_UDFS["prep_method_step"]
+        prep_method: str | None = self._get_methods(step_names_udfs, lims_id)
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:247]: Use named args

File: cg/apps/lims/api.py [hunk 11]
```diff
@@ -239,19 +241,17 @@ def get_sample_attribute(self, lims_id: str, key: str) -> str:
...
+        return prep_method
 
-        return self._get_methods(step_names_udfs, lims_id)
-
-    def get_sequencing_method(self, lims_id: str) -> str:
+    def get_sequencing_method(self, lims_id: str) -> str | None:
         """Get the sequencing method."""
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:251]:
```suggestion
        """Return the sequencing method."""
```

File: cg/apps/lims/api.py [hunk 12]
```diff
@@ -239,19 +241,17 @@ def get_sample_attribute(self, lims_id: str, key: str) -> str:
...
-
         step_names_udfs = MASTER_STEPS_UDFS["sequencing_method_step"]
-
-        return self._get_methods(step_names_udfs, lims_id)
+        sequencing_method: str | None = self._get_methods(step_names_udfs, lims_id)
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:253]: Use named args

File: cg/apps/lims/api.py [hunk 13]
```diff
@@ -266,54 +266,55 @@ def _sort_by_date_run(sort_list: list):
         """
         return sorted(sort_list, key=lambda sort_tuple: sort_tuple[0], reverse=True)
 
-    def _get_methods(self, step_names_udfs, lims_id):
+    def _get_methods(self, step_names_udfs, lims_id) -> str | None:
         """
         Gets the method, method number and method version for a given list of step names for AM documents.
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:271]:
```suggestion
        Return the method, method number and method version for a given list of step names for AM documents.
```

File: cg/apps/lims/api.py [hunk 14]
```diff
@@ -266,54 +266,55 @@ def _sort_by_date_run(sort_list: list):
...
-                )
-                methods.append(
-                    (
-                        process.date_run,
-                        self.get_method_document(process, udf_key_method_doc),
-                        self.get_method_version(process, udf_key_version),
-                        method_type,
+        try:
+            for process_name in step_names_udfs:
+                artifacts = self.get_artifacts(process_type=process_name, samplelimsid=lims_id)
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:277]: Type hint

File: cg/apps/lims/api.py [hunk 15]
```diff
@@ -266,54 +266,55 @@ def _sort_by_date_run(sort_list: list):
...
+                if not artifacts:
+                    continue
+                # Get a list of parent processes for the artifacts
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:280]:
```suggestion
```

File: cg/apps/lims/api.py [hunk 16]
```diff
@@ -266,54 +266,55 @@ def _sort_by_date_run(sort_list: list):
...
+                processes: list[Process] = self.get_processes_from_artifacts(artifacts=artifacts)
+                for process in processes:
+                    # Check which type of method document has been used
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:283]:
```suggestion
```

File: cg/apps/lims/api.py [hunk 17]
```diff
@@ -266,54 +266,55 @@ def _sort_by_date_run(sort_list: list):
...
+                    method_type: str = self.get_method_type(process, step_names_udfs[process_name])
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:284]: Use named args

File: cg/apps/lims/api.py [hunk 18]
```diff
@@ -266,54 +266,55 @@ def _sort_by_date_run(sort_list: list):
...
-            elif (
-                method[METHOD_TYPE_INDEX] == DocumentationMethod.ATLAS
-                and method[METHOD_DOCUMENT_INDEX] is not None
-            ):
-                return f"{method[METHOD_DOCUMENT_INDEX]} ({method[METHOD_VERSION_INDEX]})"
+            if sorted_methods:
+                method = sorted_methods[METHOD_INDEX]
 
+                if (
+                    method[METHOD_TYPE_INDEX] == DocumentationMethod.AM
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:303]: @AnnaLeinfelt @Karl-Svard Do we still have anything in AM?

File: cg/apps/lims/api.py [hunk 19]
```diff
@@ -395,30 +396,38 @@ def get_sample_project(self, sample_id: str) -> str:
...
         return self.sample(sample_id).get("project").get("id")
 
-    def get_sample_rin(self, sample_id: str) -> float:
+    def get_sample_rin(self, sample_id: str) -> float | None:
         """Return the sample RIN value."""
-        sample_artifact: Artifact = Artifact(self, id=f"{sample_id}PA1")
-        return sample_artifact.udf.get(PROP2UDF["rin"])
+        rin = None
+        try:
+            sample_artifact: Artifact = Artifact(self, id=f"{sample_id}PA1")
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:403]:
```suggestion
            sample_artifact = Artifact(self, id=f"{sample_id}PA1")
```

Comment (henrikstranneheim) [reply cg/apps/lims/api.py:403]: @Karl-Svard What is the `PA1` constant? Can we make it a named variable?

File: cg/apps/lims/api.py [hunk 20]
```diff
@@ -395,30 +396,38 @@ def get_sample_project(self, sample_id: str) -> str:
...
+                for artifact in artifacts:
+                    input_amounts.append(
+                        (
+                            artifact.parent_process.date_run,
+                            artifact.udf.get(udf_key),
+                        )
                     )
-                )
+        except HTTPError as error:
+            LOG.warning(f"Could not fetch sample {sample_id} from LIMS: {error}")
```
Comment (henrikstranneheim) [inline cg/apps/lims/api.py:430]: Can we make these more specific and also include the reason for failing?

File: cg/meta/report/report_api.py [hunk 21]
```diff
@@ -305,7 +296,9 @@ def get_workflow_accreditation_limitation(self, application_tag: str) -> str | N
         )
         return application_limitation.limitations if application_limitation else None
 
-    def get_sample_application_data(self, sample: Sample, lims_sample: dict) -> ApplicationModel:
+    def get_sample_application_data(
```
Comment (henrikstranneheim) [inline cg/meta/report/report_api.py:299]:
```suggestion
    def get_sample_application(
```

File: cg/meta/report/report_api.py [hunk 22]
```diff
@@ -305,7 +296,9 @@ def get_workflow_accreditation_limitation(self, application_tag: str) -> str | N
...
+        self, sample: Sample, lims_sample: dict[str:Any]
+    ) -> ApplicationModel:
         """Retrieves the analysis application attributes."""
```
Comment (henrikstranneheim) [inline cg/meta/report/report_api.py:302]:
```suggestion
        """Return the analysis application attributes."""
```

File: cg/meta/upload/scout/scout_config_builder.py [hunk 23]
```diff
@@ -158,13 +148,13 @@ def include_sample_alignment_file(self, config_sample: ScoutIndividual) -> None:
...
     def get_sample_file(self, hk_tags: set[str], sample_id: str) -> str | None:
-        """Return a file that is specific for a individual from housekeeper"""
+        """Return a file that is specific for an individual from Housekeeper."""
         tags: set = hk_tags.copy()
         tags.add(sample_id)
         return self.get_file_from_hk(hk_tags=tags)
 
     def get_file_from_hk(self, hk_tags: set[str], latest: bool | None = False) -> str | None:
-        """Get a file from housekeeper and return the path as a string."""
+        """Get a file from Housekeeper and return the path as a string."""
```
Comment (henrikstranneheim) [inline cg/meta/upload/scout/scout_config_builder.py:157]:
```suggestion
        """Return the Housekeeper file path as a string."""
```

File: cg/meta/workflow/analysis.py [hunk 24]
```diff
@@ -473,10 +474,17 @@ def get_lims_naming_metadata(self, sample: Sample) -> str | None:
         return None
 
     def get_latest_metadata(self, case_id: str) -> AnalysisModel:
-        """Get the latest metadata of a specific case"""
+        """Get the latest analysis metadata of a specific case."""
```
Comment (henrikstranneheim) [inline cg/meta/workflow/analysis.py:477]:
```suggestion
        """Return the latest analysis metadata of a specific case."""
```

File: cg/meta/workflow/balsamic.py [hunk 25]
```diff
@@ -625,3 +626,42 @@ def report_deliver(self, case_id: str, dry_run: bool = False) -> None:
...
+    def get_variant_callers(self, analysis_metadata: BalsamicAnalysis) -> list[str]:
+        """
+        Return list of Balsamic variant-calling filters and their versions (if available) from the
+        config.json file.
+        """
+        sequencing_type: str = analysis_metadata.config.analysis.sequencing_type
+        analysis_type: str = analysis_metadata.config.analysis.analysis_type
+        var_callers: dict[str, BalsamicVarCaller] = analysis_metadata.config.vcf
+        tool_versions: dict[str, list] = analysis_metadata.config.bioinfo_tools_version
+        analysis_var_callers = list()
```
Comment (henrikstranneheim) [inline cg/meta/workflow/balsamic.py:651]:
```suggestion
        analysis_var_callers = []
```

**Review by henrikstranneheim — APPROVED** [review]
Great cleanup, 👍

File: cg/apps/lims/api.py [hunk 26]
```diff
@@ -395,30 +396,38 @@ def get_sample_project(self, sample_id: str) -> str:
         """Get the lims-id for the project of the sample"""
```
Comment (diitaz93) [inline cg/apps/lims/api.py:396]: Update docstring and return type hint to account when the sample does not exist. Same for method above
```suggestion
        """Get the lims-id for the project of the sample, if the sample exists."""
```

**Review by diitaz93 — APPROVED** [review]
Looks great 💯

File: cg/apps/lims/api.py (hunk 1 above)
Comment (ivadym) [reply cg/apps/lims/api.py:57]: We're using the sample method and the LIMS sample object interchangeably, so I'd prefer to handle it in another PR 😓

File: cg/apps/lims/api.py (hunk 3 above)
Comment (ivadym) [reply cg/apps/lims/api.py:137]: Same here, it's probably safer to address it in a separate PR and typehint all the calls

File: cg/apps/lims/api.py (hunk 5 above)
Comment (ivadym) [reply cg/apps/lims/api.py:143]: https://github.com/Clinical-Genomics/cg/issues/3163

File: cg/apps/lims/api.py (hunk 19 above)
Comment (ivadym) [reply cg/apps/lims/api.py:403]: Added it to the issue for future refactoring: https://github.com/Clinical-Genomics/cg/issues/3163

File: cg/apps/lims/api.py (hunk 18 above)
Comment (ivadym) [reply cg/apps/lims/api.py:303]: Added it to the issue for future refactoring: https://github.com/Clinical-Genomics/cg/issues/3163

File: cg/apps/lims/api.py (hunk 26 above)
Comment (ivadym) [reply cg/apps/lims/api.py]: Good catch ⭐ 

[31093b1](https://github.com/Clinical-Genomics/cg/pull/3153/commits/31093b1d65805f780abe76b89b07a07b90a3679c)

File: tests/meta/report/conftest.py [hunk 27]
```diff
@@ -75,9 +76,9 @@ def case_samples_data(case_id: str, report_api_mip_dna: MipDNAReportAPI):
 
 
 @pytest.fixture(scope="function")
-def mip_analysis_api(cg_context: CGConfig) -> MockMipAnalysis:
-    """MIP analysis mock data."""
-    return MockMipAnalysis(config=cg_context, workflow=Workflow.MIP_DNA)
+def mip_analysis_api(cg_context: CGConfig) -> MockMipDNAAnalysisAPI:
```
[the remark this answers is not in the document]
Comment (ivadym) [reply tests/meta/report/conftest.py:79]: 👍

**Comment by ivadym** [comment]
Deployed

**Pull request merged**